### PR TITLE
Add xo CLI auth

### PR DESCRIPTION
**Pull request by dplumb94**
### Setup apache proxy: 
`% sudo apt-get update`
`% sudo apt-get install -y apache2`
`% sudo a2enmod headers`
`% sudo a2enmod proxy_http`

`% vi /tmp/.password`
```
# .password should match this:

sawtooth:$apr1$cyAIkitu$Cv6M2hHJlNgnVvKbUdlFr.

#EOF
```

`% sudo vi /etc/apache2/sites-enabled/000-default.conf`
```
# 000-default.conf should match this:

<VirtualHost *:80>
        ServerAdmin sawtooth@sawtooth
        DocumentRoot /var/www/html

        ErrorLog ${APACHE_LOG_DIR}/error.log
        CustomLog ${APACHE_LOG_DIR}/access.log combined

        <Location />
                Options Indexes FollowSymLinks
                AllowOverride None
                AuthType Basic
                AuthName "Enter password"
                AuthUserFile "/tmp/.password"
                Require user sawtooth
                Require all denied
        </Location>
</VirtualHost>

ProxyPass /sawtooth http://localhost:8080
ProxyPassReverse /sawtooth http://localhost:8080
RequestHeader set X-Forwarded-Path "/sawtooth"

#EOF
```
`% sudo apachectl start`

### Play xo
To connect to the Apache proxy auth server, initialize xo with this command:
`% xo init --url http://localhost/sawtooth`

You can enter in auth info for each transaction with the `--auth_user` and `--auth_password` options.

Signed-off-by: Darian Plumb <dplumb@bitwise.io>

### CODE REVIEW

File: sdk/examples/xo_python/sawtooth_xo/xo_client.py [hunk 1]
```diff
@@ -103,15 +107,18 @@ def _send_request(self, suffix, data=None, content_type=None):
         else:
             url = "http://{}/{}".format(self._base_url, suffix)
 
-        headers = None
+        b64_string = b64encode(self._user.encode()).decode()
+        auth_header = 'Basic {}'.format(b64_string)
+        headers = {'Authorization': auth_header}
```
Comment (vaporos) [inline sdk/examples/xo_python/sawtooth_xo/xo_client.py]: The Authorization header should only be added if '--user' has been given.

File: sdk/examples/xo_python/sawtooth_xo/xo_cli.py [hunk 2]
```diff
@@ -170,10 +176,16 @@ def do_init(args, config):
     if args.url is not None:
         url = args.url
 
+    user = config.get('DEFAULT', 'user')
+    if args.user is not None:
+        user = args.user
```
Comment (vaporos) [inline sdk/examples/xo_python/sawtooth_xo/xo_cli.py]: At this point 'user' is the complete auth header information "USERNAME:[PASSWORD]" not just the user.  Suggest immediately breaking this into auth_user and auth_password here for clarity.

If auth_user is specified and no auth_password is specified, the user should be prompted every time xo commands are run for the password, with getpass.getpass().

File: sdk/examples/xo_python/sawtooth_xo/xo_cli.py [hunk 3]
```diff
@@ -170,10 +176,16 @@ def do_init(args, config):
...
+
     config.set('DEFAULT', 'username', username)
     print("set username: {}".format(username))
     config.set('DEFAULT', 'url', url)
     print("set url: {}".format(url))
+    config.set('DEFAULT', 'user', user)
```
Comment (vaporos) [inline sdk/examples/xo_python/sawtooth_xo/xo_cli.py]: We should move the --user argument to all sub-commands instead of during init to avoid persisting a plaintext password.

File: sdk/examples/xo_python/sawtooth_xo/xo_client.py [hunk 4]
```diff
@@ -57,7 +59,9 @@ def take(self, name, space):
     def list(self):
         xo_prefix = self._get_prefix()
 
-        result = self._send_request("state?address={}".format(xo_prefix))
+        result = self._send_request(
+            "state?address={}".format(xo_prefix)
+        )
```
Comment (delventhalz) [inline sdk/examples/xo_python/sawtooth_xo/xo_client.py]: What is the purpose of this change? It should probably be removed.

File: sdk/examples/xo_python/sawtooth_xo/xo_client.py [hunk 5]
```diff
@@ -35,9 +36,10 @@ def _sha512(data):
 
 
 class XoClient:
-    def __init__(self, base_url, keyfile):
+    def __init__(self, base_url, keyfile, user):
```
Comment (delventhalz) [inline sdk/examples/xo_python/sawtooth_xo/xo_client.py]: What will the value of `user` be if not specified in the config file? Should this parameter default o `None`?

File: sdk/examples/xo_python/sawtooth_xo/xo_client.py [hunk 6]
```diff
@@ -103,15 +107,18 @@ def _send_request(self, suffix, data=None, content_type=None):
...
             url = "http://{}/{}".format(self._base_url, suffix)
 
-        headers = None
+        b64_string = b64encode(self._user.encode()).decode()
+        auth_header = 'Basic {}'.format(b64_string)
+        headers = {'Authorization': auth_header}
+
         if content_type is not None:
-            headers = {'Content-Type': content_type}
+            headers['Content-Type'] = content_type
```
Comment (delventhalz) [inline sdk/examples/xo_python/sawtooth_xo/xo_client.py]: Use this pattern for the user authorization too:
```
headers = None

if self._user is not None:
    b64_string = b64encode(self._user.encode()).decode()
    auth_header = 'Basic {}'.format(b64_string)
    headers['Authorization'] self._auth_header
```

File: sdk/examples/xo_python/sawtooth_xo/xo_cli.py [hunk 7]
```diff
@@ -298,9 +332,17 @@ def do_take(args, config):
     url = config.get('DEFAULT', 'url')
     key_file = config.get('DEFAULT', 'key_file')
 
-    client = XoClient(base_url=url,
-                      keyfile=key_file)
-    response = client.take(name, space)
+    auth_user = args.auth_user
+    auth_password = args.auth_password
+    if auth_user is not None and auth_password is None:
+        auth_password = getpass.getpass(prompt="auth_password: ")
```
Comment (delventhalz) [inline sdk/examples/xo_python/sawtooth_xo/xo_cli.py]: You use these four lines four times in this file. Would probably be worth abstracting them into a function.

File: sdk/examples/xo_python/sawtooth_xo/xo_client.py [hunk 8]
```diff
@@ -97,21 +109,30 @@ def _get_address(self, name):
...
         else:
             url = "http://{}/{}".format(self._base_url, suffix)
 
-        headers = None
+        headers = {}
+        if auth_user is not None:
+            auth_string = "{}:{}".format(auth_user, auth_password)
+            b64_string = b64encode(auth_string.encode()).decode()
+            auth_header = 'Basic {}'.format(b64_string)
+            print(auth_header)
```
Comment (delventhalz) [inline sdk/examples/xo_python/sawtooth_xo/xo_client.py]: print statement

File: sdk/examples/xo_python/sawtooth_xo/xo_cli.py [hunk 9]
```diff
@@ -139,6 +139,16 @@ def create_parent_parser(prog_name):
         action='count',
         help='enable more verbose output')
 
+    parent_parser.add_argument(
+        '--auth_user',
+        type=str,
+        help='username to authorize request')
```
Comment (delventhalz) [inline sdk/examples/xo_python/sawtooth_xo/xo_cli.py]: This help text should probably be clear this is to support Basic Auth if the REST API is using it.

File: sdk/examples/xo_python/sawtooth_xo/xo_cli.py [hunk 10]
```diff
@@ -139,6 +139,16 @@ def create_parent_parser(prog_name):
...
+
+    parent_parser.add_argument(
+        '--auth_password',
+        type=str,
+        help='password to authorize request')
```
Comment (delventhalz) [inline sdk/examples/xo_python/sawtooth_xo/xo_cli.py]: Same. This is only for if the REST API is using the Basic Auth authorization scheme.

File: sdk/examples/xo_python/sawtooth_xo/xo_cli.py [hunk 11]
```diff
@@ -139,6 +139,16 @@ def create_parent_parser(prog_name):
         action='count',
         help='enable more verbose output')
 
+    parent_parser.add_argument(
+        '--auth_user',
```
Comment (vaporos) [inline sdk/examples/xo_python/sawtooth_xo/xo_cli.py]: Underscores are not appropriate in command line arguments - use --auth-user and --auth-password instead.

File: sdk/examples/xo_python/sawtooth_xo/xo_cli.py (hunk 7 above)
Comment (vaporos) [inline sdk/examples/xo_python/sawtooth_xo/xo_cli.py]: "Auth Password: " would be friendlier.

File: sdk/examples/xo_python/sawtooth_xo/xo_cli.py [hunk 12]
```diff
@@ -139,6 +139,16 @@ def create_parent_parser(prog_name):
         action='count',
         help='enable more verbose output')
 
+    parent_parser.add_argument(
+        '--auth-user',
+        type=str,
+        help='username for authentication if Rest API is using Basic Auth')
```
Comment (delventhalz) [inline sdk/examples/xo_python/sawtooth_xo/xo_cli.py]: REST API

File: sdk/examples/xo_python/sawtooth_xo/xo_cli.py [hunk 13]
```diff
@@ -139,6 +139,16 @@ def create_parent_parser(prog_name):
...
+
+    parent_parser.add_argument(
+        '--auth-password',
+        type=str,
+        help='password for authentication if Rest API is using Basic Auth')
```
Comment (delventhalz) [inline sdk/examples/xo_python/sawtooth_xo/xo_cli.py]: REST API

File: sdk/examples/xo_python/sawtooth_xo/xo_cli.py [hunk 14]
```diff
@@ -242,9 +254,14 @@ def do_show(args, config):
...
     key_file = config.get('DEFAULT', 'key_file')
+    auth_user, auth_password = get_auth_info(args)
 
     client = XoClient(base_url=url, keyfile=key_file)
-    data = client.show(name)
+
+    data = client.show(
+        name,
+        auth_user=auth_user,
+        auth_password=auth_password)
```
Comment (delventhalz) [inline sdk/examples/xo_python/sawtooth_xo/xo_cli.py]: This could be one line (78 chars total).

File: sdk/examples/xo_python/sawtooth_xo/xo_cli.py [hunk 15]
```diff
@@ -337,6 +364,15 @@ def save_config(config):
     os.rename("{}.new".format(config_file), config_file)
 
 
+def get_auth_info(args):
```
Comment (delventhalz) [inline sdk/examples/xo_python/sawtooth_xo/xo_cli.py]: This should be underscored since it is a private function: `_get_auth_info`. If underscored, PyLint will prevent it being used outside this module.

File: sdk/examples/xo_python/sawtooth_xo/xo_client.py [hunk 16]
```diff
@@ -69,22 +78,24 @@ def list(self):
...
             return None
 
-    def _get_status(self, batch_id, wait):
+    def _get_status(self, batch_id, wait, auth_user=None, auth_password=None):
         try:
             result = self._send_request(
-                'batch_status?id={}&wait={}'.format(batch_id, wait)
-            )
+                'batch_status?id={}&wait={}'.format(batch_id, wait),
+                auth_user=None,
```
Comment (peterschwarz) [inline sdk/examples/xo_python/sawtooth_xo/xo_client.py]: This should be passing the `auth_user`, `auth_password` from the args to this call.
